src/Order.java

```diff
@@ -0,0 +1,96 @@
+import java.math.BigDecimal;
+
+public class Order {
+
+    private Client purchaser;
+    private PaymentMethod paymentMethod;
+    private Products product;
+    private Integer quantity;
+    private BigDecimal estimatedTotal;
+
+
+    public Order() {
+    }
+
+    public Order(Client purchaser, Products product, Integer quantity, PaymentMethod paymentMethod) {
+        this.purchaser = purchaser;
+        this.product = product;
+        this.quantity = quantity;
+        this.paymentMethod = paymentMethod;
+    }
+
+
+    public Client getPurchaser() {
+        return this.purchaser;
+    }
+
+    public PaymentMethod getPaymentMethod() {
+        return this.paymentMethod;
+    }
+
+    public Products getProduct() {
+        return this.product;
+    }
+
+    public Integer getQuantity() {
+        return quantity;
+    }
+
+    public BigDecimal getEstimatedTotal() {
+            return estimatedTotal;
+    }
+
+
+    public void setPurchaser(Client purchaser) {
+        this.purchaser = purchaser;
+    }
+
+    public void setPaymentMethod(PaymentMethod paymentMethod) {
+        this.paymentMethod = paymentMethod;
+    }
+
+    public void setProduct(Products product) {
+        this.product = product;
+    }
+
+    public void setQuantity(Integer quantity) {
+        this.quantity = quantity;
+    }
+
+
+    public boolean decreaseQuantity(){
+        if (this.product.getQuantityProduct() >= this.quantity){
+            this.product.setQuantityProduct(this.product.getQuantityProduct()-this.quantity);
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    public void setEstimatedTotal () {
+        try{
+        if (this.decreaseQuantity()) {
+            this.estimatedTotal = this.getProduct().getPrice().multiply(new BigDecimal(this.getQuantity()));
+            this.paymentMethod.setValue(this.estimatedTotal);
+        }
+        else {
+            System.out.println("quantidade de produto insuficiênte");
+        }
+        }
+        catch(NullPointerException exception){
+            System.out.println("O preço do produto não está cadastrado");
+        }
+    }
```

Comment:
Sempre será o preço do produto o problema?
Faz sentido este System.out aqui?

Reply:
Então, aqui (setEstimatedTotal) e no discountEstimatedTotal eu havia colocado esse "tratamento" porque lá na classe produto eu defini que o preço não pode ser menor do que zero. Sendo assim, quando a pessoa tenta cadastrar um produto com valor negativo, ele avisa a pessoa que deu um erro no cadastro mas o produto é criado e o valor atribuído ao preço é null. Quando a pessoa tenta comprar esse produto com preço null dá problema aqui. Agora eu penso em duas opções, ou fazer um tratamento mais genérico aqui usando throw new Exception, porque realmente podem ocorrer outros erros, ou criar um if que verifica se o preço é diferente de null ai ele executa, se não, dá uma mensagem.
if (this.getProduct().getPrice() != null){
executa;
} else {
sout ("O produto está sem preço cadastrado);
}

Reply:
Faça o seguinte teste na sua classe principal (main)

`Order order = new Order();
order.setEstimatedTotal();`

Coloque apenas isto e veja se faz sentido...

Reply:
Então, quando eu faço isso ocorre:
Exception in thread "main" java.lang.NullPointerException
	at Order.decreaseQuantity(Order.java:62)
	at Order.setEstimatedTotal(Order.java:72)
	at Main.main(Main.java:46)

É uma NullPointerException, não tenho que tratar?

Reply:
Posso colocar um throws PriceException na assinatura do método?

Reply:
Acho que o ponto que o @gilsonexacta está comentando é que, observe a diferença do metodo setEstimatedTotal() para outros setters da sua classe.

Tem algo de estranho?